File: 3A/se382/A4/SpriteCanvas.java
/*
 *
 */
import java.awt.Color;
import java.awt.Graphics;
import java.awt.Graphics2D;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.KeyEvent;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.awt.event.MouseMotionAdapter;
import java.awt.event.MouseWheelEvent;
import java.awt.event.MouseWheelListener;
import java.io.EOFException;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.Serializable;
import java.util.Iterator;
import java.util.Vector;

import javax.swing.BoxLayout;
import javax.swing.JFrame;
import javax.swing.JLabel;
import javax.swing.JMenu;
import javax.swing.JMenuBar;
import javax.swing.JMenuItem;
import javax.swing.JOptionPane;
import javax.swing.JPanel;
import javax.swing.KeyStroke;


/**
 * A canvas that draws sprites and provides capabilities for recording
 * events and playing them back.
 * 
 * Michael Terry
 */
@SuppressWarnings("serial")
public class SpriteCanvas extends JPanel {

    /**
     * Tracks our current interactive mode
     */
    private enum InteractionMode {
        IDLE,
        RECORDING,
        PLAYBACK
    }

    private Vector<Sprite>          sprites             = new Vector<Sprite>(); // All sprites we're managing
    private Sprite                  interactiveSprite   = null;                 // Sprite with which user is interacting
    private InteractionMode         interactionMode     = InteractionMode.IDLE; // Current interactive mode
    private Vector<Serializable>    eventStream         = null;                 // Event stream for recording events
    private long                    lastTime            = -1;                   // Time of last event
    
    public SpriteCanvas() {
        initialize();
    }

    private void initialize() {
        // Install our event handlers
        this.addMouseListener(new MouseAdapter() {
            public void mousePressed(java.awt.event.MouseEvent e) {
                handleMousePress(e);
            }
            public void mouseReleased(MouseEvent e) {
                handleMouseReleased(e);
            }
        });
        this.addMouseMotionListener(new MouseMotionAdapter() {
            public void mouseDragged(MouseEvent e) {
                handleMouseDragged(e);
            }
        });
        this.addMouseWheelListener(new MouseWheelListener() {
          public void mouseWheelMoved(MouseWheelEvent e) {
            handleMouseWheelMoved(e);
          }
        });
    }
    
    /**
     * Handle mouse press events from user or demo file
     */
    private void handleMouseWheelMoved(java.awt.event.MouseWheelEvent e) {
      for (Sprite sprite : sprites) {
          interactiveSprite = sprite.getSpriteHit(e);
          if (interactiveSprite != null) {
              recordEvent(e);
              interactiveSprite.handleMouseWheelMovedEvent(e);
              break;
          }
      }
      repaint();
    }
    
    private void handleMousePress(java.awt.event.MouseEvent e) {
        for (Sprite sprite : sprites) {
            interactiveSprite = sprite.getSpriteHit(e);
            if (interactiveSprite != null) {
                recordEvent(e);
                interactiveSprite.handleMouseDownEvent(e);
                break;
            }
        }
    }
    
    /**
     * Handle mouse released events from user or demo file
     */
    private void handleMouseReleased(MouseEvent e) {
        if (interactiveSprite != null) {
            recordEvent(e);
            interactiveSprite.handleMouseUp(e);
            repaint();
        }
        interactiveSprite = null;
    }
    
    /**
     * Handle mouse dragged events from user or demo file
     */
    private void handleMouseDragged(MouseEvent e) {
        if (interactiveSprite != null) {
            recordEvent(e);
            interactiveSprite.handleMouseDragEvent(e);
            repaint();
        }
    }

    /**
     * Record an event, but only if recording is turned on
     */
    private void recordEvent(MouseEvent event) {
        if (interactionMode.equals(InteractionMode.RECORDING)) {
            long thisTime = System.currentTimeMillis();
            eventStream.add(thisTime - lastTime);
            eventStream.add(event);
            lastTime = thisTime;
        }
    }
    
    /**
     * Add a top-level sprite to manage
     */
    public void addSprite(Sprite s) {
        sprites.add(s);
    }
    
    /**
     * Paint our canvas
     */
    public void paint(Graphics g) {
        g.setColor(Color.WHITE);
        g.fillRect(0, 0, this.getWidth(), this.getHeight());
        g.setColor(Color.BLACK);
        for (Sprite sprite : sprites) {
            sprite.draw((Graphics2D)g);
        }
        paintChildren(g);
    }
    
    /**
     * Starts recording of events.
     */
    public void startRecording() {
        interactionMode = InteractionMode.RECORDING;
        eventStream = new Vector<Serializable>();
        lastTime = System.currentTimeMillis();
    }
    
    /**
     * Stops recording events, writes them to a file called "demo"
     */
    public void stopRecording() {
        interactionMode = InteractionMode.IDLE;
        if (eventStream != null) {
            try {
                ObjectOutputStream out = new ObjectOutputStream(new FileOutputStream("demo"));
                for (Serializable item : eventStream) {
                    out.writeObject(item);
                }
                out.close();
            } catch (FileNotFoundException e) {
                e.printStackTrace();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }
    
    /**
     * Starts the demo. Important note: This simply plays back events
     * that were previously recorded. Thus, it is assumed that the sprites
     * are in the *exact same state* as when recording started. If not, you
     * will get unexpected results. This demo code DOES NOT try to reset
     * the sprites to the initial state they were in when recording started.
     * You may want to add this feature.
     * 
     * A message box will be shown if the demo file cannot be found.
     */
    public void startDemo() {
        interactionMode = InteractionMode.PLAYBACK;
        final Vector<Serializable> playbackStream = new Vector<Serializable>();
        try {
            ObjectInputStream in = new ObjectInputStream(new FileInputStream("demo"));
            while (true) {
                Serializable event = (Serializable)in.readObject();
                playbackStream.add(event);
            }
        } catch (FileNotFoundException e) {
            JOptionPane.showMessageDialog(this, "Demo file not found.");
        } catch (EOFException e) {
            ; // ignore: This is how we know we're done reading the file
        } catch (IOException e) {
            e.printStackTrace();
        } catch (ClassNotFoundException e) {
            e.printStackTrace();
        }
        
        // Create and start a thread to play back the events
        Thread playbackThread = new Thread(new Runnable() {
            public void run() {
                boolean hadFirstWait = false;
                Iterator<Serializable> iter = playbackStream.iterator();
                while (iter.hasNext() && interactionMode.equals(InteractionMode.PLAYBACK)) {
                    Long waitTime = (Long)iter.next();
                    MouseEvent event = (MouseEvent)iter.next();
                    try {
                        if (hadFirstWait) {
                            Thread.sleep(waitTime);
                        } else {
                            hadFirstWait = true;
                        }
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                    switch (event.getID()) {
                        case MouseEvent.MOUSE_PRESSED:
                            handleMousePress(event);
                            break;
                        case MouseEvent.MOUSE_DRAGGED:
                            handleMouseDragged(event);
                            break;
                        case MouseEvent.MOUSE_RELEASED:
                            handleMouseReleased(event);
                            break;
                    }
                }
            }
        });
        playbackThread.start();
    }
    
    /**
     * Stops the demo.
     */
    public void stopDemo() {
        interactionMode = InteractionMode.IDLE;
    }

    private static Sprite buildPaperDoll() {
      Sprite body = new Body();
      body.translate(350,150);
      
      Sprite head = new Head();
      head.translate(80, 0);
      Sprite lleg = new LLeg();
      lleg.translate(40,195);
      Sprite rleg = new RLeg();
      rleg.translate(115,195);
      Sprite larm = new LArm();
      larm.translate(10, 40);
      Sprite rarm = new RArm();
      rarm.translate(150, 40);
      body.addChild(head);
      body.addChild(lleg);
      body.addChild(rleg);
      body.addChild(larm);
      body.addChild(rarm);
      
      Sprite llowleg = new LLowLeg();
      llowleg.translate(0,95);
      Sprite rlowleg = new RLowLeg();
      rlowleg.translate(0,95);
      Sprite llowarm = new LLowArm();
      llowarm.translate(-10,45);
      Sprite rlowarm = new RLowArm();
      rlowarm.translate(10,45);
      lleg.addChild(llowleg);
      rleg.addChild(rlowleg);
      larm.addChild(llowarm);
      rarm.addChild(rlowarm);
      
      Sprite lfoot = new LFoot();
      lfoot.translate(-5,85);
      Sprite rfoot = new RFoot();
      rfoot.translate(5,85);
      Sprite lhand = new LHand();
      lhand.translate(-5,75);
      Sprite rhand = new RHand();
      rhand.translate(5,75);
      llowleg.addChild(lfoot);
      rlowleg.addChild(rfoot);
      llowarm.addChild(lhand);
      rlowarm.addChild(rhand);
      
      // Link together the legs for scaling
      lleg.addScaleBuddy(rleg);
      lleg.addScaleBuddy(llowleg);
      lleg.addScaleBuddy(rlowleg);
      rleg.addScaleBuddy(lleg);
      rleg.addScaleBuddy(llowleg);
      rleg.addScaleBuddy(rlowleg);
      llowleg.addScaleBuddy(lleg);
      llowleg.addScaleBuddy(rleg);
      llowleg.addScaleBuddy(rlowleg);
      rlowleg.addScaleBuddy(lleg);
      rlowleg.addScaleBuddy(rleg);
      rlowleg.addScaleBuddy(llowleg);
      
      return body;
    }
    
    private JMenuBar buildMenuBar() {
      JMenuBar bar = new JMenuBar();
      JMenu menu;
      JMenuItem item;
      
      menu = new JMenu("File");
      item = new JMenuItem("Reset");
      item.setAccelerator(KeyStroke.getKeyStroke(KeyEvent.VK_R, ActionEvent.CTRL_MASK));
      item.addActionListener(new ActionListener() {
        public void actionPerformed(ActionEvent arg0) {
          sprites = new Vector<Sprite>();
          addSprite(buildPaperDoll());
          repaint();
        }
      });
      menu.add(item);
      menu.addSeparator();
      
      item = new JMenuItem("Quit");
      item.addActionListener(new ActionListener() {
        public void actionPerformed(ActionEvent arg0) {
          System.exit(0);
        }
      });
      menu.add(item);
      
      bar.add(menu);
      
      return bar;
    }
    
    /**
     * Test code.
     */
    public static void main(String[] args) {
        Sprite body = buildPaperDoll();
      
        SpriteCanvas canvas = new SpriteCanvas();
        canvas.addSprite(body);
        
        JFrame f = new JFrame();
        f.setJMenuBar(canvas.buildMenuBar());
        f.getContentPane().setLayout(new BoxLayout(f.getContentPane(), BoxLayout.Y_AXIS));
        canvas.add(new JLabel("Click to move and rotate.  Use the scroll wheel to scale the legs."));
        f.getContentPane().add(canvas);
        f.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        f.setSize(900, 650);
        f.setVisible(true);
    }

}

File: 3A/se382/A5/SpriteCanvas.py
import java.awt.Color as Color
import java.awt.Graphics as Graphics
import java.awt.Graphics2D as Graphics2D
import java.awt.Rectangle as Rectangle
import java.awt.event.ActionEvent as ActionEvent
import java.awt.event.ActionListener as ActionListener
import java.awt.event.KeyEvent as KeyEvent
import java.awt.event.MouseAdapter as MouseAdapter
import java.awt.event.MouseEvent as MouseEvent
import java.awt.event.MouseMotionAdapter as MouseMotionAdapter
import java.awt.event.MouseWheelEvent as MouseWheelEvent
import java.awt.event.MouseWheelListener as MouseWheelListener
import java.io.EOFException as EOFException
import java.io.FileInputStream as FileInputStream
import java.io.FileNotFoundException as FileNotFoundException
import java.io.FileOutputStream as FileOutputStream
import java.io.IOException as IOException
import java.io.ObjectInputStream as ObjectInputStream
import java.io.ObjectOutputStream as ObjectOutputStream
import java.io.Serializable as Serializable
import java.util.Iterator as Iterator
import java.util.Vector as Vector

import javax.swing.BoxLayout as BoxLayout
import javax.swing.ImageIcon as ImageIcon
import javax.swing.JFrame as JFrame
import javax.swing.JLabel as JLabel
import javax.swing.JMenu as JMenu
import javax.swing.JMenuBar as JMenuBar
import javax.swing.JMenuItem as JMenuItem
import javax.swing.JOptionPane as JOptionPane
import javax.swing.JPanel as JPanel
import javax.swing.KeyStroke as KeyStroke

import java.lang.Math as Math
import java.lang.System as System
from Sprite import Sprite
from Sprite import DRAGGING
from Sprite import ROTATING

class SpriteCanvas(JPanel):
  def __init__(self):
    self.interactiveSprite = None
    self.sprites = []
    self.initialize()

  def addSprite(self, s):
    self.sprites.append(s)

  def clearSprites(self):
    self.sprites = []

  #
  # Inner Classes
  #
  class scMouseAdapter(MouseAdapter):
    def __init__(self, canvas):
      self.canvas = canvas
    def mousePressed(self, e):
      canvas.handleMousePress(e)
    def mouseReleased(self, e):
      canvas.handleMouseReleased(e)

  class scMouseMotionAdapter(MouseMotionAdapter):
    def __init__(self, canvas):
      self.canvas = canvas
    def mouseDragged(self, e):
      canvas.handleMouseDragged(e)

  class scMouseWheelListener(MouseWheelListener):
    def __init__(self, canvas):
      self.canvas = canvas
    def mouseWheelMoved(self, e):
      canvas.handleMouseWheelMoved(e)

  class scResetListener(ActionListener):
    def __init__(self, canvas):
      self.canvas = canvas
    def actionPerformed(self, e):
      canvas.clearSprites()
      canvas.addSprite(buildPaperDoll())
      canvas.repaint()

  class scQuitListener(ActionListener):
    def __init__(self, canvas):
      self.canvas = canvas
    def actionPerformed(self, e):
      System.exit(0)

  # Initialize the canvas
  def initialize(self):
    self.addMouseListener(self.scMouseAdapter(self))
    self.addMouseMotionListener(self.scMouseMotionAdapter(self))
    self.addMouseWheelListener(self.scMouseWheelListener(self))

  # Handler Methods
  def handleMousePress(self, e):
    for sprite in self.sprites:
      self.interactiveSprite = sprite.getSpriteHit(e)
      if self.interactiveSprite is not None:
        self.interactiveSprite.handleMouseDownEvent(e)
        break
    self.repaint()

  def handleMouseReleased(self, e):
    for sprite in self.sprites:
      self.interactiveSprite = sprite.getSpriteHit(e)
      if self.interactiveSprite is not None:
        self.interactiveSprite.handleMouseUp(e)
        break
    self.repaint()

  def handleMouseDragged(self, e):
    if self.interactiveSprite is not None:
      self.interactiveSprite.handleMouseDragEvent(e)
      self.repaint()

  def handleMouseWheelMoved(self, e):
    for sprite in self.sprites:
      self.interactiveSprite = sprite.getSpriteHit(e)
      if self.interactiveSprite is not None:
        self.interactiveSprite.handleMouseWheelMovedEvent(e)
        break
    self.repaint()


  # Drawing
  def paint(self, g):
    g.setColor(Color.WHITE)
    g.fillRect(0, 0, self.getWidth(), self.getHeight())
    g.setColor(Color.BLACK)
    for sprite in self.sprites:
      sprite.draw(g)
    self.paintChildren(g)

  def buildMenuBar(self):
    bar = JMenuBar()
    menu = JMenu()
    item = JMenuItem()

    menu = JMenu("File")
    item = JMenuItem("Reset")
    item.setAccelerator(KeyStroke.getKeyStroke(KeyEvent.VK_R, ActionEvent.CTRL_MASK))
    item.addActionListener(self.scResetListener(self))
    menu.add(item)
    menu.addSeparator()

    item = JMenuItem("Quit")
    item.addActionListener(self.scQuitListener(self))
    menu.add(item)

    bar.add(menu)

    return bar

def buildPaperDoll():
  body = Sprite(Rectangle(0,0,160,220), 0,
                ImageIcon("body.png"), 350, 150,
                DRAGGING, False)

  head = Sprite(Rectangle(-55,-80,110,110), Math.PI * (5.0/18),
                ImageIcon("head.png"), 80, 0,
                ROTATING, False, body)
  lleg = Sprite(Rectangle(-20,-20,40,110), Math.PI * 0.5,
                ImageIcon("leg.png"), 40, 195,
                ROTATING, True, body)
  rleg = Sprite(Rectangle(-20,-20,40,110), Math.PI * 0.5,
                ImageIcon("leg.png"), 115, 195,
                ROTATING, True, body)
  larm = Sprite(Rectangle(-25,-30,40,90), Math.PI,
                ImageIcon("up_arm.png"), 10, 40,
                ROTATING, False, body)
  rarm = Sprite(Rectangle(-15,-30,40,90), Math.PI,
                ImageIcon("r_up_arm.png"), 150, 40,
                ROTATING, False, body)

  rlowleg = Sprite(Rectangle(-20,-20,40,110), Math.PI * 0.5,
                   ImageIcon("r_low_leg.png"), 0, 95,
                   ROTATING, True, rleg)
  llowleg = Sprite(Rectangle(-20,-20,40,110), Math.PI * 0.5,
                   ImageIcon("low_leg.png"), 0, 95,
                   ROTATING, True, lleg)
  llowarm = Sprite(Rectangle(-20,0,40,90), Math.PI * (135.0/180),
                   ImageIcon("low_arm.png"), -10, 45,
                   ROTATING, False, larm)
  rlowarm = Sprite(Rectangle(-20,0,40,90), Math.PI * (135.0/180),
                   ImageIcon("r_low_arm.png"), 10, 45,
                   ROTATING, False, rarm)

  lfoot = Sprite(Rectangle(-70,-10,90,65), Math.PI * (35.0/180),
                 ImageIcon("foot.png"), -5, 85,
                 ROTATING, False, llowleg)
  rfoot = Sprite(Rectangle(-20,-10,90,65), Math.PI * (35.0/180),
                 ImageIcon("r_foot.png"), 5, 85,
                 ROTATING, False, rlowleg)
  lhand = Sprite(Rectangle(-20,5,35,65), Math.PI * (35.0/180),
                 ImageIcon("hand.png"), -5, 75,
                 ROTATING, False, llowarm)
  rhand = Sprite(Rectangle(-15,5,35,65), Math.PI * (35.0/180),
                 ImageIcon("r_hand.png"), 5, 75,
                 ROTATING, False, rlowarm)

  lleg.addScaleBuddy(rleg)
  lleg.addScaleBuddy(llowleg)
  lleg.addScaleBuddy(rlowleg)
  rleg.addScaleBuddy(lleg)
  rleg.addScaleBuddy(llowleg)
  rleg.addScaleBuddy(rlowleg)
  llowleg.addScaleBuddy(lleg)
  llowleg.addScaleBuddy(rleg)
  llowleg.addScaleBuddy(rlowleg)
  rlowleg.addScaleBuddy(lleg)
  rlowleg.addScaleBuddy(rleg)
  rlowleg.addScaleBuddy(llowleg)

  return body


if __name__ == "__main__":
  f = JFrame()
  body = buildPaperDoll()
  canvas = SpriteCanvas()
  canvas.addSprite(body)

  f.setJMenuBar(canvas.buildMenuBar())
  f.getContentPane().setLayout(BoxLayout(f.getContentPane(), BoxLayout.Y_AXIS))
  canvas.add(JLabel("Click and drag to move and rotate.  Use the scroll wheel to scale the legs."))
  f.getContentPane().add(canvas)
  f.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE)
  f.setSize(900, 650)
  f.setVisible(True)

